jwt_proxy/policy_engine.py

```diff
@@ -95,14 +117,158 @@ def evaluate_policies(req, user_info=None):
+                )
+        except Exception as e:
+            current_app.logger.exception("Request transformer %s raised exception: %s", file_path, e)
+            continue
+    
+    return modified_body
+
+
+def apply_response_transformers(req, response_body, user_info=None):
+    """Apply all response transformers to modify the response body.
+    
+    Transformers should have signature: transform_response(request, response_body, user_info) -> modified_response_body
+    Returns the modified response body, or None if a FHIR resource was filtered out.
+    
+    Note: Transformers can return:
+    - Modified body (dict): use this as the new body
+    - None: indicates no modification needed OR FHIR resource was filtered
+      - Caller should check if original was FHIR resource to determine if filtered
+    """
+    from flask import current_app
+    
+    if not isinstance(response_body, dict):
+        return response_body
+    
+    # Track if we started with a FHIR resource
+    is_fhir_resource = response_body.get("resourceType") is not None
+    
+    modified_body = response_body
+    filtered = False
+    
+    for transformer, file_path in _loaded_response_transformers:
+        transformer_name = os.path.basename(file_path)
+        current_app.logger.info(
+            "Applying response transformer: %s for %s %s",
+            transformer_name,
+            req.method,
+            req.path or req.url if hasattr(req, 'url') else 'unknown'
+        )
+        
+        # Capture the body before transformation for comparison
+        body_before = json.dumps(modified_body, sort_keys=True) if modified_body else None
+        
+        try:
+            result = transformer(req, modified_body, user_info)
+            if result is not None:
+                body_after = json.dumps(result, sort_keys=True) if result else None
+                if body_before != body_after:
+                    current_app.logger.info(
+                        "Response transformer %s modified the response body",
+                        transformer_name
+                    )
+                else:
+                    current_app.logger.debug(
+                        "Response transformer %s returned result but made no changes",
+                        transformer_name
+                    )
+                # Transformer returned a modified body
+                modified_body = result
+                filtered = False  # Reset filtered flag if we got a result
+            elif is_fhir_resource and isinstance(modified_body, dict) and modified_body.get("resourceType"):
+                # Transformer returned None for a FHIR resource - it was filtered
+                current_app.logger.info(
+                    "Response transformer %s filtered out FHIR resource",
+                    transformer_name
+                )
+                filtered = True
+                break  # Stop processing, resource was filtered
```

Comment:
@ivan-c This seems to be causing issues for the /fhir/Patient?identifier=http://keycloak.cirg.uw.edu|{id} requests, and likely any other searches - the elif block catches all transformer results that are None (which seems to generally mean "ignore" elsewhere) and instead treats them as "deny" if the request body is a fhir resource, which it is for any search bundle result. Any clear reason for this?